File: frontend/src/models/dish.ts
//
// Copyright (c) 2013-2016 SKYARCH NETWORKS INC.
//
// This software is released under the MIT License.
//
// http://opensource.org/licenses/mit-license.php
//

/// <reference path="../../declares.d.ts" />

import ModelBase from './base';

export default class Dish extends ModelBase {
  constructor(public id: string) {super(); }

  static ajax = new AjaxSet.Resources('dishes');


  runlist(id: number): JQueryPromise<any> {
    return this.WrapAndResolveReject(() =>
      (<any>Dish.ajax).runlist({id: id})
    );
  }
}
Dish.ajax.add_member('runlist', 'GET');

File: app/models/dish.rb
#
# Copyright (c) 2013-2017 SKYARCH NETWORKS INC.
#
# This software is released under the MIT License.
#
# http://opensource.org/licenses/mit-license.php
#

class Dish < ActiveRecord::Base
  belongs_to :project
  has_many :resources
  has_many :dish_servertests
  has_many :servertests, through: :dish_servertests

  serialize :runlist

  STATUS = {
    success: 'SUCCESS',
    failure: 'FAILURE',
    creating: 'CREATING',
    bootstrapping: 'BOOTSTRAPPING',
    applying: 'APPLYING',
    serverspec: 'SERVERSPEC',
  }.recursive_freeze

  def update_status(symbol_status)
    self.status = STATUS[symbol_status]
    save!
  end

  def validating?
    case status
    when STATUS[:creating], STATUS[:bootstrapping], STATUS[:applying], STATUS[:serverspec]
      return true
    end

    false
  end

  def playbook_roles_safe
    return [] if playbook_roles.nil?
    JSON.parse(playbook_roles)
  end

  def extra_vars_safe
    if extra_vars.nil?
      return '{}'
    end
    extra_vars
  end

  def self.valid_dishes(project_id = nil)
    where(project_id: [nil, project_id]).where(status: STATUS[:success])
  end
end
